Run each e2e test function in a separate namespace

sbar95 opened the pull request:
<!--
    Please read https://github.com/KohlsTechnology/eunomia/blob/master/.github/CONTRIBUTING.md before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

<!-- Please provide a summary of the change here. -->
Run each test function in separate namespace.

Add SetupRbacInNamespace function to make it possible to deploy a
service account to on-the-fly generated namespace (during tests).

In each test deploy eunomia-operator service account bound to
eunomia-operator role. Also remove underscores from test function names
to make per-namespace tests possible because test namespaces created
dynamically by operator-sdk framework take function name as part of a
namespace (and such cannot have an underscore).

Make it possible to use TEMPLATE_PROCESSOR_ARGS env variable in helm
template processor.

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #211 

**Type of change**

<!-- Please delete options that are not relevant. -->

* New feature (non-breaking change which adds functionality)

**Checklist**

- [x] Unit tests and e2e tests updated
- [ ] Documentation updated (no need for this PR)

akavel commented on template-processors/helm/bin/processTemplates.sh:
```diff
@@ -24,4 +24,5 @@ helm init --client-only
 echo "Generating manifest files"
-helm template -f /tmp/eunomia_values_processed.yaml --output-dir "${MANIFEST_DIR}" --namespace "${NAMESPACE}" "${CLONED_TEMPLATE_GIT_DIR}"
+helm template -f /tmp/eunomia_values_processed.yaml ${TEMPLATE_PROCESSOR_ARGS:-} \
+  --output-dir "${MANIFEST_DIR}" --namespace "${NAMESPACE}" "${CLONED_TEMPLATE_GIT_DIR}"
```
1. In the commit message that introduces the changes here, could you please add a reference to the issue where TEMPLATE_PROCESSOR_ARGS were originally introduced? This could make it easier to understand the bigger context of this change 
2. [optional] What would you think about splitting this command such that each flag+parameter is on a separate line?

akavel commented on deploy/test/eunomia-rbac/templates/role-operator.yaml:
```diff
@@ -0,0 +1,55 @@
+  - eunomia-operator
+  verbs:
+  - "update"
+{{- end }}
```
How does this file differ from the already existing file `deploy/helm/eunomia-operator/templates/role-operator.yaml`? why can't we reuse that existing one and we have to create a new very similar (or is it identical?) file?

seanmalloy replied:
I don't think we should have special RBAC that is only deployed when running the tests. We should just you our standard RBAC configs when running the tests.

sbar95 replied:
This is the standard test RBAC, but per namespace. Until now it was being deployed each time to the [sole test namespace `test-eunomia-operator`](https://github.com/KohlsTechnology/eunomia/blob/master/scripts/e2e-test.sh#L52) with the use of [`helm template`](https://github.com/KohlsTechnology/eunomia/blob/master/scripts/e2e-test.sh#L50-L52):
- https://github.com/KohlsTechnology/eunomia/blob/master/deploy/helm/eunomia-operator/templates/role-operator.yaml
- https://github.com/KohlsTechnology/eunomia/blob/master/deploy/helm/eunomia-operator/templates/service-account-operator.yaml

We had these in only one test namespace (because all tests were run in only one). Now with each test function being run in a separate test namespace I need the Role, ServiceAccount, and RoleBinding in each namespace for the tests not to raise access errors.

sbar95 replied:
@akavel I'll try to reuse the already existing files

sbar95 replied:
They are now being reused: 22073da.
Later, I'll squash it and edit this message too.

akavel commented on test/e2e/util.go:
```diff
@@ -1,4 +1,4 @@
-// +build e2e
+// +build e2e xd
```
What's the `xd` here for?

seanmalloy commented on scripts/e2e-test.sh:
```diff
@@ -51,7 +51,14 @@ helm template deploy/helm/eunomia-operator/ \
+  -test.parallel=1 \
+  -localOperator \
+  -tags e2e \
+  -v
```
I'm curious on this one too. Can we ask the `operator-sdk` maintainers if there is a way to make this work? Maybe we can open an issue or post in their slack channel?

sbar95 replied:
I cannot run `operator-sdk test local`'s `--up-local` flag without the `--namespace` flag:
https://github.com/operator-framework/operator-sdk/blob/v0.8.1/cmd/operator-sdk/test/local.go#L116-L117 (it's also the case in the latest operator-sdk).
<details><summary><code>operator-sdk test local --help</code></summary><br><pre>
$ operator-sdk test local --help
Run End-To-End tests locally

Usage:
  operator-sdk test local <path to tests directory> [flags]

Flags:
      --debug                        Enable debug-level logging
      --global-manifest string       Path to manifest for Global resources (e.g. CRD manifests)
      --go-test-flags string         Additional flags to pass to go test
  -h, --help                         help for local
      --image string                 Use a different operator image from the one specified in the namespaced manifest
      --kubeconfig string            Kubeconfig path
      --molecule-test-flags string   Additional flags to pass to molecule test
      --namespace string             If non-empty, single namespace to run tests in
      --namespaced-manifest string   Path to manifest for per-test, namespaced resources (e.g. RBAC and Operator manifest)
      --no-setup                     Disable test resource creation
      --up-local                     Enable running operator locally with go run instead of as an image in the cluster

Global Flags:
      --verbose   Enable verbose logging
</pre></details>
However, it's still possible to run the tests successfully by issuing `operator-sdk test local ./test/e2e --no-setup --verbose --go-test-flags "-tags e2e -timeout 20m"` (deploying eunomia to the cluster beforehand) **but** then all the tests are being run in the `default` namespace (and that's because of [GetNamespace() function](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/pkg/test/resource_creator.go#L36-L39) - `*singleNamespace` is in turn enforced by [`--no-setup` flag](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/cmd/operator-sdk/test/local.go#L205-L207)).

**Finally**
Well, in course of writing this comment I actually found the right `oprator-sdk`'s flag combination to run the tests successfully in separate namespaces (no need to resort to bare `go test ...`), but eunomia has to be deployed to the cluster beforehand (I cannot use `--up-local`) but I think this is fine, because apart from e2e tests we now have a deployment test at the same time (so we don't need a [separate test](https://github.com/KohlsTechnology/eunomia/blob/master/scripts/e2e-test.sh#L60-L80) at the end of the `e2e-test.sh` script - and if the deployment fails, e2e tests just won't run):
```
operator-sdk test local ./test/e2e \
--namespaced-manifest /dev/null \
--global-manifest /dev/null \
--verbose \
--go-test-flags "-tags e2e -timeout 20m"
```
Soooo... Thank you for the comments, they helped me a lot! 👍

sbar95 commented on test/e2e/util.go:
```diff
@@ -189,3 +194,43 @@ func debugJSON(v interface{}) string {
+		"helm", "template",
+		filepath.Join(eunomiaRoot, "deploy/test/eunomia-rbac/"),
+		"--set", "eunomia.namespace="+namespace,
+	).Output()
```
There is the [InitializeClusterResources() function](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/pkg/test/resource_creator.go#L108) using namespacedMan parameter I didn't see earlier - I'll try to come up with something using it (but I will also need a namespaced manifest dir - maybe I'll be generating it dynamically somehow).

sbar95 replied:
The function I linked to isn't enough unfortunately. I can create a single manifest file like this one:
<details><summary><code>my_manifest.yml</code></summary><br><pre>
---
# Source: eunomia-rbac/templates/service-account-operator.yaml
# service account for the operator
apiVersion: v1
kind: ServiceAccount
metadata:
  name: eunomia-operator
  namespace: eunomia-operator
---
# Source: eunomia-rbac/templates/role-operator.yaml
kind: Role
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: eunomia-operator
  namespace: "eunomia-operator"
rules:
# leader election   
- apiGroups:
  - ""
  resources:
  - configmaps
  - pods
  verbs:
  - "*"  
#Metrics  
- apiGroups:
  - ""
  resources:
  - services
  verbs:
  - "*"
#Metrics  
- apiGroups:
  - "apps"
  resources:
  - replicasets
  - deployments
  verbs: 
  - "*"
# In Kubernetes <1.16, "deployments" also require API group "extensions" (see: https://stackoverflow.com/a/57854939)
- apiGroups:
  - "extensions"
  resources:
  - deployments
  verbs:
  - "*"
# Metrics  
- apiGroups:
  - monitoring.coreos.com
  resources:
  - servicemonitors
  verbs:
  - "get"
  - "create"
# Metrics  
- apiGroups:
  - apps
  resources:
  - deployments/finalizers
  resourceNames:
  - eunomia-operator
  verbs:
  - "update"
---
# Source: eunomia-rbac/templates/service-account-operator.yaml
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: eunomia-operator
  namespace: eunomia-operator
subjects: 
- kind: ServiceAccount
  name: eunomia-operator
  namespace: eunomia-operator
roleRef:
  kind: Role
  name: eunomia-operator
  apiGroup: rbac.authorization.k8s.io
</pre></details>

and pass it to operator-sdk: (`operator-sdk test local ./test/e2e --namespaced-manifest /path/to/my_manifest.yml`.

But this is what happens:
In each test function I invoke:
- [`InitializeClusterResources`](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/pkg/test/resource_creator.go#L108),
- which returns [`ctx.createFromYAML`](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/pkg/test/resource_creator.go#L55),
- this function [sets the current namespace](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/pkg/test/resource_creator.go#L72) in the resources provided by my `my_manifest.yml`,
- but unfortunately only in the [metadata section](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/unstructured/unstructured.go#L237-L243).

This is a problem for me because I need the namespace to be also set in `my_manifest.yml` in RoleBinding's subjects subsection. It doesn't happen and the namespace there stays set to the initial value (it isn't being changed per test to the dynamically created namespace as in the metadata section) so I get access errors.

sbar95 replied:
After additional research I found out that if you don't pass `--namespaced-manifest` flag with content to the `operator-sdk test local` command (and at the same time you don't provide `--no-setup` flag) the program looks for [some default files](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/cmd/operator-sdk/test/local.go#L122) which we do not have in `deploy/` directory (maybe they were created by the `operator-sdk` initially but then removed at some point in time because we started using helm?).

sbar95 commented:
@seanmalloy @akavel 
Please, re-review. I pushed some additional commits addressing the comments. I also put some answers.
Please, if it's acceptable, don't merge for now - I'm going to squash my commits then and remove the WIP label.

akavel commented on test/e2e/util.go:
```diff
@@ -189,3 +194,57 @@ func debugJSON(v interface{}) string {
+		"--set", "eunomia.operator.deployment.nsRbacOnly=true",
+	).Output()
+	if err != nil {
+		return fmt.Errorf("Failed to generate manifest files")
```
Please don't lose the value present in `err` - please use xerrors.Errorf + %w format:

    return xerrors.Errorf("Failed to generate manifest files: %w", err)

— please also fix **in all other places in this func**

akavel commented on test/e2e/util.go:
```diff
@@ -189,3 +194,57 @@ func debugJSON(v interface{}) string {
+	eunomiaRoot := filepath.Join(filepath.Dir(utilTestFilename), "../..")
+
+	// create temporary file to store kubectl input in
+	tmpFile, err := ioutil.TempFile(os.TempDir(), "sa-")
```
please leave first arg to TempFile empty `""` (*["If dir is the empty string, TempFile uses the default directory for temporary files"](https://golang.org/pkg/io/ioutil/#TempFile)*)

akavel commented on test/e2e/util.go:
```diff
@@ -189,3 +194,57 @@ func debugJSON(v interface{}) string {
+	}
+
+	// create role, role binding and service account
+	cmd := exec.Command("kubectl", "apply", "-f", tmpFile.Name())
```
instead of creating a temporary file on disk, I believe we should be able to just pass the `out` contents as in-memory stdin to kubectl:

```go
cmd := exec.Command(..., "-f", "-")
cmd.Stdin = bytes.NewReader(out)
```

seanmalloy commented on scripts/e2e-test.sh:
```diff
@@ -47,34 +47,32 @@ fi
-
+# Eunomia setup
 helm template deploy/helm/eunomia-operator/ \
   --set eunomia.operator.image.tag=dev \
```
So ... I think this means we are actually running the container image from quay.io instead of the container image built from the local branch.

sbar95 replied:
Not in this case - in build-images.sh [freshly built images are always being tagged `dev`](https://github.com/KohlsTechnology/eunomia/blob/bab5f717b69c6599183ad4d9f7ac833859f14e74/scripts/build-images.sh#L43-L44).
I just wanted to be explicit here - when somebody reads this test script, he'll associate the `dev` tag with the development eunomia image (which he might've just modified).
Besides, the image can't be pulled from quay because of the `pullPolicy` set to `Never` one line below.

seanmalloy replied:
Ok ... sounds good.

seanmalloy commented:
@sbar95 this is now a merge conflict with `scripts/e2e-test.sh` that you will need to resolve.

sbar95 commented on deploy/helm/eunomia-operator/templates/deployment.yaml:
```diff
@@ -1,5 +1,5 @@
 {{- with .Values.eunomia.operator }}
-{{- if .deployment.enabled -}}
+{{- if and (.deployment.enabled) (not .deployment.nsRbacOnly) -}}
```
When initially I was using bare `go test` instead of `operator-sdk test local` I tried to use the `-test.parallel=4` flag in one run to see what happens but there was no difference in test time - we need `t.Parallel()`.

To enable parallelism in `operator-sdk test local` we can use the `--go-test-flags` flag to pass the option. The command by itself sets the parallel flag in [only one case](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/cmd/operator-sdk/test/local.go#L205-L207) (not ours, fortunately).

In the end `operator-sdk test local` is just [a wrapper for `go test`](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/cmd/operator-sdk/test/local.go#L211-L222) - [it just uses exec](https://github.com/operator-framework/operator-sdk/blob/v0.8.1/internal/util/projutil/exec.go#L76-L86).
